python/hail/utils/java.py

```diff
@@ -186,39 +187,78 @@ def get():
             Log4jLogger.log_pkg = Env.jutils()
         return Log4jLogger.log_pkg
 
+
 def error(msg):
     Log4jLogger.get().error(msg)
 
+
 def warn(msg):
     Log4jLogger.get().warn(msg)
 
+
 def info(msg):
     Log4jLogger.get().info(msg)
 
-@decorator
-def handle_py4j(func, *args, **kwargs):
-
-    try:
-        r = func(*args, **kwargs)
-    except py4j.protocol.Py4JJavaError as e:
-        tpl = Env.jutils().handleForPython(e.java_exception)
-        deepest, full = tpl._1(), tpl._2()
-        raise FatalError('%s\n\nJava stack trace:\n%s\n'
-                         'Hail version: %s\n'
-                         'Error summary: %s' % (deepest, full, Env.hc().version, deepest))
-    except py4j.protocol.Py4JError as e:
-        if e.args[0].startswith('An error occurred while calling'):
-            msg = 'An error occurred while calling into JVM, probably due to invalid parameter types.'
-            raise FatalError('%s\n\nJava stack trace:\n%s\n'
-                             'Hail version: %s\n'
-                             'Error summary: %s' % (msg, e.message, Env.hc().version, msg))
-        else:
-            raise e
-    except CapturedException as e:
-        raise FatalError('%s\n\nJava stack trace:\n%s\n'
-                         'Hail version: %s\n'
-                         'Error summary: %s' % (e.desc, e.stackTrace, Env.hc().version, e.desc))
-    return r
+
+def handle_java_exception(f):
+    def deco(*args, **kwargs):
+        _exception = None
+        try:
+            return f(*args, **kwargs)
+        except py4j.protocol.Py4JJavaError as e:
+            s = e.java_exception.toString()
+
+            # py4j catches NoSuchElementExceptions to stop array iteration
+            if s.startswith('java.util.NoSuchElementException'):
+                raise
+
+            tpl = Env.jutils().handleForPython(e.java_exception)
+            deepest, full = tpl._1(), tpl._2()
+            _exception = FatalError('%s\n\nJava stack trace:\n%s\n'
```

Comment:
these error messages repeat the first part of the message again under "Error summary". Was that intentional? It always kind of confused me. Can we get rid of one of them or replace one with a more succinct message?

Reply:
(Maybe I'm just getting confused and the part that was actually confusing was that the old stack trace used to get printed, which you fixed.)